tensorflow/compiler/mlir/xla/transforms/hlo_outline_fusion.cc

```diff
@@ -0,0 +1,555 @@
+        if (!cluster_to) {
+          continue;
+        }
+
+        bool contracted_edge = fn(cluster_from, cluster_to);
+        changed |= contracted_edge;
+      }
+    }
+
+    return changed;
+  }
+
+  std::vector<Value> GetResultsOfFusedPattern(Cluster* from, Cluster* to) {
+    auto fused_pattern = MergeFusionPattern(
+        from->fused_pattern(), to->fused_pattern());
+    return GetOutputsOfFusionPattern(fused_pattern);
+  }
+
+  bool TryToContractEdge(Cluster* from, Cluster* to) {
+    int node_to = to->cycles_graph_node_id();
+    int node_from = from->cycles_graph_node_id();
+
+    // Both node_to and node_from should be fusible
+    if (!IsFusible(op_list_[node_to]) || !IsFusible(op_list_[node_from])) {
+      return false;
+    }
+
+    // ReduceOp is fusible only when it's the output of the fusion pattern.
+    if (dyn_cast<xla_hlo::ReduceOp>(op_list_[node_from])) {
+      return false;
+    }
+
+    // Only Try to fuse along data edge, not index edge
+    if (IsShapeProducer(op_list_[node_from])) {
+      return false;
+    }
+
+    // All outputs of a kLoop fusion should be same.
+    auto results = GetResultsOfFusedPattern(from, to);
+    auto get_workload_shape = [] (Value v) {
+      if (dyn_cast<xla_hlo::ReduceOp>(v.getDefiningOp())) {
+        v = v.getDefiningOp()->getOperand(0);
+      }
+      return v;
+    };
+    if (!llvm::all_of(results, [&] (Value result) {
+      Value ref = get_workload_shape(results[0]);
+      Value val = get_workload_shape(result);
+      return shape_analysis_.HasSameShape(ref, val);
+    })) {
+      return false;
+    }
+
+    return MergeClusters(from, to);
+  }
+
+  bool RunEdgeContractionLoop() {
+    bool changed = false;
+    changed = ForEachEdgeInPostOrder([&](Cluster* from, Cluster* to) -> bool {
+      return TryToContractEdge(from, to);
+    });
+    return changed;
+  }
+
+  SmallVector<Operation*, 4> op_list_;
+
+  // Shape equality checker
+  ShapeConstrainAnalysis shape_analysis_;
+
+  std::unordered_map<Operation*, int> op_to_node_id_;
+
+  GraphCycles cycle_detector_;
+  std::vector<std::unique_ptr<Cluster>> cluster_storage_;
+  std::vector<UnionFind<Cluster*>> cluster_for_node_;
+};
+
+struct HloOutlineFusion : public ModulePass<HloOutlineFusion> {
+  void runOnModule() override {
+    SmallVector<FuncOp, 4> funcs;
+    // We may add some fusion func into the module during
+    // visiting, thus we collect all target funcs first.
+    for (FuncOp func : getModule().getOps<FuncOp>()) {
+      if (IsTargetFunc(func)) {
+        funcs.push_back(func);
+      }
+    }
+    for (FuncOp func : funcs) {
+      runOnFunction(func);
+    }
+  }
+
+  bool IsTargetFunc(FuncOp func) {
+    // We currently only support single block function.
+    if (func.getBlocks().size() != 1) {
+      return false;
+    }
+
+    if (llvm::none_of(func.getBlocks().front(), [](Operation& op) {
+        return (op.getDialect()->getNamespace() == "xla_hlo" ||
```

Comment:
This is expensive. Use `context->getRegisteredDialect<"xla_hlo">()` and the compare the `Dialect` objects. Those are unique within the context.